### ISSUE TITLE
hhvm occasionally hangs on average load

### ISSUE THREAD

**Issue by genych** [facebook/hhvm#8029]
Vague problem maybe related to #5081

We have a VM with Ubuntu 16.04, HHVM-3.21.3 (same results on 3.22.1) and Nginx configured according to hhvm manuals.

Under low load it works just fine, but in a specific situation hhvm/hh_server (belonging to the web server user) permanently stop responding and causes nginx timeout. No high load, nothing special, just a lot of curl waiting, some mysql, some IO. No frameworks, almost no async. Basically a long sequence of send_request→wait→process→store→send_another + a bit of web site interaction in the background.

Sometimes hhvm process hangs with 100% CPU usage but most of the time there is nothing unusual: bunch of sleeping hh_server processes from users and web server, bunch of sleeping hhvm processes. 

We tried manually built hhvm 3.21.3 instead and so far it works without issues, but I'm uncertain because it's kinda hard to reproduce and test.

What can it be? What can we try? What additional info will be helpful to debug this weird issue?

**Comment by genych** [comment]
Another test fails with
```
[err] event_queue_remove: 0x7f69061982d8(fd 9) not on queue 8
[err] event.c:782: Assertion TAILQ_EMPTY(&base->eventqueue) failed in event_base_free
Core dumped: Aborted
Stack trace in /tmp/stacktrace.2143.log
Aborted (core dumped)
```
Do you need trace/dump/perf.map?

**Comment by genych** [comment]
It also fails with
```
pure virtual method called
terminate called without an active exception
Core dumped: Aborted
Stack trace in /tmp/stacktrace.16671.log
Aborted (core dumped)
```

**Comment by usox** [comment]
I can confirm this (somewhat older) issue. Having a couple of sequential curl-requests which take a long time (up to the default timeout of 60s) results in hhvms webserver getting unresponsive.
Will try to provide a code sample to reproduce the behaviour...

**Comment by usox** [comment]
Ok. I pinpoint my issue which was caused by the hhvm thread-settings which was way too low in my setup.

**Comment by fredemmott** [comment]
Closing this as:
 - the crashes/hangs where information is provided here are a KP, fixed by 3.24.8 and 3.27.1+
 - the curl busy loop is being separately tracked
